### DISCUSSION TITLE
Record overrides are overridden

### DISCUSSION THREAD

**Discussion by digitaldias** [dotnet/csharplang#4135] [Language Ideas] [answered]
VS2019 v16.8.1
C# 9

overrides are either ignored or overridden by the compiler in records. 

**Steps to Reproduce**:
Create the following record: 
```
    public abstract record RecordOf<T>
    {
        public T Value { get; set; }

        public override string ToString()
        {
            return Value.ToString();
        }
    }
```
Create a record that extends this base record: 
`public record EmployeeCapacity : RecordOf<int>{};`

**Expected Behavior**:
I have a restaurant record that has a property defined as 
`public EmployeeCapacity ChefCapacity{get;set;}
`
And it is output to a blazor page as such: 

`<td>@restaurant.ChefCapacity</td>
`
I expect to see for example "1". 

**Actual Behavior**:
I see the entire record, such as 
`EmployeeCapacity { Value = 1 }`

**Comment by Youssef1313** [comment]
I believe the current behavior is known to be expected. So I think this needs a language design review; or probably this scenario could be a good candidate for a source generator.

**Comment by 333fred** [comment, marked as answer]
I've converted this to a discussion on csharplang, as the record feature is behaving exactly as expected here and it is not a bug. As we discussed in discord, you probably just want a `class` that can inherit from a `record`, which will hopefully be coming in C# next.

**Reply by Youssef1313** [reply]
@333fred I read through the spec, and I think there is a some sort of spec bug/confusion here.

https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/records#inheritance

The spec says:

> Members are synthesized unless a member with a "matching" signature is declared in the record body or an accessible concrete non-virtual member with a "matching" signature is inherited.

However, this doesn't seem to apply to the synthesized `ToString` (there is always a matching signature inherited from System.Object)

**Reply by CyrusNajmabadi** [reply]
> there is always a matching signature inherited from System.Object

That signature in Object is virtual though. So this doesn't apply.

**Reply by Youssef1313** [reply]
@CyrusNajmabadi Yep I missed that, but what about something like:

```csharp
public record A
{
    public override string ToString()
    {
        return "Record A.";
    }
}

public record B : A;
```

`A.ToString` is an `override` not `virtual`. However, ToString is still synthesized in record B.

**Reply by HaloFour** [reply]
`override`s are `virtual` unless `sealed`.

**Reply by Youssef1313** [reply]
@HaloFour That makes sense. Thanks!

**Reply by CyrusNajmabadi** [reply]
Yup. Abstract/override/virtual are all in the 'virtual' family.

**Reply by digitaldias** [reply]
> `override`s are `virtual` unless `sealed`.

Well, that makes it hard to create ValueTypes out of records, which is one huge benefit of records that I was hoping for. 

```
    public record Money : RecordOf<decimal>
    {

    }

    public void DepositMoney(Money amount)
    {
        // impl.
    }
```
Because outputting things like the Amount to a logger now implies I have to explicitly call `amount.Value` since there is no way in texas I can override the tostring().

**Reply by canton7** [reply]
If the LDT wanted to support this case, would it be acceptable to let users seal the `ToString` method (or any other method)? This currently results in:

> error CS8870: 'RecordOf<T>.ToString()' cannot be sealed because containing record is not sealed.
> error CS0239: 'Money.ToString()': cannot override inherited member 'RecordOf<int>.ToString()' because it is sealed

But I'd personally find it intuitive as a way of telling the compiler "back off, I've got this".

**Reply by jnm2** [reply]
@canton7 Fred is championing this: https://github.com/dotnet/csharplang/issues/4174

**Reply by canton7** [reply]
Oh! Very nice, thanks.

**Reply by canton7** [reply]
Suggest writing that as an answer and accepting it? I think that solves this issue entirely.

**Reply by digitaldias** [reply]
Super!! That will solve my problem! Thumbs up!
